File: src/main/java/more/arithmetic/AM4.java
package more.arithmetic;

import java.util.HashMap;

/**
 * @ClassName AM4
 * @Description
 *
 * 编写一个函数来查找字符串数组中的最长公共前缀。
 *
 * 如果不存在公共前缀，返回空字符串 ""。
 *
 * 示例 1:
 *
 * 输入: ["flower","flow","flight"]
 * 输出: "fl"
 * 示例 2:
 *
 * 输入: ["dog","racecar","car"]
 * 输出: ""
 * 解释: 输入不存在公共前缀。
 * 说明:
 *
 * 所有输入只包含小写字母 a-z 。
 *
 * @Author root
 * @Date 18-11-7 上午9:17
 * @Version 1.0
 **/
public class AM4 {

    public static void main (String args[]) {

        String[] strs = {"aca","cba"};

        System.out.println(longestCommonPrefix(strs));
    }

    /**
     *
     * 个人解法  执行用时: 17 ms, 在Longest Common Prefix的Java提交中击败了26.56% 的用户
     * 功能描述: 取出字符数组中,最短的一个字符串,进行循环比较,
     *
     * @param:
     * @return:
     * @auther: Fmbah
     * @date: 18-11-7 上午9:19
     */
    public static String longestCommonPrefix(String[] strs) {

        String result = "";
        if (strs.length > 0) {
            String minlenStr = strs[0];
            for (String str : strs) {
                if (str.length() < minlenStr.length()) {
                    minlenStr = str;
                }
            }

            HashMap hashMap = new HashMap();
            for (int i = 0; i < minlenStr.toCharArray().length; i++) {
                hashMap.put(i, minlenStr.toCharArray()[i] + "");
            }

            for (int i= 0, j = hashMap.keySet().size(); i < j; i++) {
                String str = (String)hashMap.get(i);
                boolean canContract = true;
                for (int x = 0, y = strs.length; x < y; x++) {
                    String tmp = strs[x].toCharArray()[i] + "";
                    if (!tmp.equals(str)) {
                        canContract = false;
                        break;
                    }
                }
                if (canContract) {
                    result += str;
                } else if (!canContract && i == 0) {
                    break;
                }
            }
        }

        return result;
    }

}

File: src/main/java/more/arithmetic/AM4.py
class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        使用zip函数,进行字符串对接
        执行用时: 28 ms, 在Longest Common Prefix的Python提交中击败了93.96% 的用户
        """
        if strs is None or len(strs) == 0:
            return ''
        for i in range(len(strs)):
            strs[i] = list(strs[i])
        tmp = zip(*strs)
        res = ''
        for i in tmp:
            if len(set(i)) == 1:
                res += i[0]
            else:
                break
        return res

    def longestCommonPrefix1(self, strs):
        """
        两两比较
        @param: strs: A list of strings
        @return: The longest common prefix
        """
        if strs is None or len(strs) == 0:
            return ''
        res = strs[0]
        for i in range(1, len(strs)):
            tmp = res
            res = ''
            for j in range(min(len(strs[i]),len(tmp))):
                if tmp[j] == strs[i][j]:
                    res += tmp[j]
                else:
                    break
        return res

s = Solution()
print(s.longestCommonPrefix1(['aca', 'acba', 'bac']))
for j in range(len('abc')):
    print(j, 'abc'[j])
